### DISCUSSION TITLE
using wLength for OS detection?

### DISCUSSION THREAD

**Discussion by thepfy** [hathach/tinyusb#1986] [Q&A]
Hi!

Do you think its possible to do OS detection on a RP2040 running tinyUSB as a HID in a similar way to how https://github.com/keyboardio/FingerprintUSBHost works?

I've been using the HID composite example from the Pico SDK and I've found the control packets in process_control_request and process_get_descriptor methods from usbd.c and have then been sending them out on UART using printf but all of the wLengths are all hex FF, which from other reading wouldn't normally be the case and I haven't been able to figure out why all the packets seem to have the same wLength but I am not a very good C programmer.....

Would something like this be possible in TinyUSB or am I barking up the wrong tree?

**Comment by hathach** [comment]
each host driver may behave slightly diffeferent, for example, windows can ask for string descriptor of 0xEF (MS OS string v1 extension something). A part from that any other way is just observation and can be incorrect in the future.
